designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,348 @@
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    defaults:
+      budgets: 
+        - nodes: 10% 
+          schedule: "0 0 1 * *"
+          duration: 1h 
+    consolidation:
+      consolidationPolicy: WhenUnderutilized
+      disruptAfter: "30m"
+    drift:
+      budgets:
+        - nodes: "20%"
+          schedule: "0 0 1 * *"
+          duration: "1h"
+        - nodes: "10%"
+          schedule: "0 0 * * 0"
+          duration: "2h"
+        - nodes: "50%" 
+          schedule: "@monthly"
+```
+#### Considerations 
+Some of the API choices for a given reason seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each reason. It appears there is a need for disruption controls within the budgets for each reason, not just overall.
+
+This approach aligns well with controls that apply to all existing reasons. The proposal presented here is similar to the one mentioned above in relation to the reasons we allow to be defined (underutilized, drifted, expired, empty).
+
+This proposal is currently scoped for disruptionBudgets by reason. However, we should also consider incorporating other generic disruption controls into the PerReasonControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-reason controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-reason level of granularity.
+
+### Pros + Cons 
+* 👍 Granular Control for Each Reason: This model aligns well with the need for specific controls for each disruption reason, providing a natural specification for further granularity. 
+* 👍 Foundation for Future Extensions: Offers a framework for incorporating other generic disruption controls into the per-reason controls, extending beyond budgets.
+* 👍👍 Customization for Specific Reasons: Allows for specific validations and extensions for individual reasons, offering tailored control.
+* 👎👎👎 Breaking API Changes:Implementing this approach would necessitate significant changes to the current API, especially for budgets. If we decide to model DisruptAfter in a similar way we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
+* 👎 Complexity in Default Reason Handling: The model does not facilitate easy defaulting for unspecified ("default") disruption reasons.
+* 👎 Increased Budget Complexity: The use of budgets becomes more complex as they are now nested within another field, adding a layer of intricacy to their application.
+
+### API Design Conclusion: Preferred Design
```

Comment:
Should this just directly reference approach A?

Reply:
It does now